File: problems/32-k-way-merge/002-smallest-range-k-lists/python/solution.py
import heapq
class Solution:
    def smallest_range(self, nums):
        heap=[]; current_max=-10**30
        for i, arr in enumerate(nums):
            heapq.heappush(heap,(arr[0],i,0)); current_max=max(current_max,arr[0])
        best=(heap[0][0],current_max)
        while len(heap)==len(nums):
            value,i,j=heapq.heappop(heap)
            if current_max-value < best[1]-best[0]: best=(value,current_max)
            j+=1
            if j==len(nums[i]): break
            nxt=nums[i][j]; heapq.heappush(heap,(nxt,i,j)); current_max=max(current_max,nxt)
        return list(best)

File: problems/32-k-way-merge/002-smallest-range-k-lists/java/Solution.java
import java.util.*;
public class Solution {
    public int[] smallestRange(List<List<Integer>> nums) {
        PriorityQueue<int[]> pq = new PriorityQueue<>(Comparator.comparingInt(a -> a[0]));
        int max = Integer.MIN_VALUE;
        for (int i=0;i<nums.size();i++) { int v=nums.get(i).get(0); pq.add(new int[]{v,i,0}); max=Math.max(max,v); }
        int bestL=pq.peek()[0], bestR=max;
        while (pq.size()==nums.size()) {
            int[] cur=pq.poll(); int min=cur[0];
            if ((long)max-min < (long)bestR-bestL) { bestL=min; bestR=max; }
            int ni=cur[2]+1;
            if (ni==nums.get(cur[1]).size()) break;
            int v=nums.get(cur[1]).get(ni); pq.add(new int[]{v,cur[1],ni}); max=Math.max(max,v);
        }
        return new int[]{bestL,bestR};
    }
}